### PR TITLE
Add password strength meter

### PR DESCRIPTION
**Pull request by orangenbaumblatt**
I tinkered around with implementing a password strength meter as mentioned in #421.

There are a few things to do:

- [x] Fix a few UI bugs and improve layout
- [x] Check if string arrays can be translated with Crowdin
- [x] Translate password feedback strings (currently a won't do)
- [x] Maybe implement a warning dialog if the password is to weak (zxcvbn offers suggestions which can be used here)
- [x] Take a look at increasing app size (~2MB) due to dictionaries

### CODE REVIEW

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -256,4 +256,12 @@
...
     <string name="empty_list">There are no codes to be shown. Start adding entries by tapping the plus sign in the bottom right corner</string>
     <string name="empty_list_title">No entries found</string>
+
+    <string-array name="password_strength">
+        <item>Weak</item>
+        <item>Fair</item>
+        <item>Good</item>
+        <item>Strong</item>
+        <item>Excellent</item>
+    </string-array>
```
[the remark this answers is not in the document]
Comment (orangenbaumblatt) [reply app/src/main/res/values/strings.xml:266]: Yes, that might be better. Do you know if these kinds of `string-arrays` can be translated with Crowdin?

[the remark this answers is not in the document]
Comment (michaelschattgen) [reply app/src/main/res/values/strings.xml:266]: We have another file 'arrays.xml' for string-arrays. You should move this string array to that file and replace the strings with references to string values. So your string array will become something like this:

```
    <string-array name="password_strength">
        <item>@string/password_strength_very_weak</item>
        <item>@string/password_strength_weak</item>
        <item>@string/password_strength_fair</item>
        <item>@string/password_strength_good</item>
        <item>@string/password_strength_streng</item>
    </string-array>
```

This way the string will become translatable.

[the remark this answers is not in the document]
Comment (orangenbaumblatt) [reply app/src/main/res/values/strings.xml]: Thanks for the tip!

**Comment by alexbakker** [comment]
This is starting to look pretty good to me!

[](https://alexbakker.me/u/3v5kvhaqtu.png)

Can you make it so that the progress bar doesn't stick out, but aligns and has the same width as the text boxes above? 

I personally don't think a warning dialog is necessary. The warning message and strength indicator are hard to miss. I'm also not too worried about the increase in APK size. I think this feature justifies it.

**Comment by orangenbaumblatt** [comment]
Sure, fixed it. Also tried making the the feedback messages translatable via `strings.xml`, but couldn't come with a solution.
zxcvbn4j uses Java ResourceBundles which don't seem to play nicely with the way Android does it. If it's too much of an issue, we can consider switching to [nbvcxz](https://github.com/GoSimpleLLC/nbvcxz), which has some translations already built in.

**Comment by alexbakker** [comment]
Does nbvcxz do translations differently? It looks the same to me.

**Comment by orangenbaumblatt** [comment]
@alexbakker Yeah, it's basically the same but with some translations already included.
I'm just looking for a nice way to translate these messages with the help of `strings.xml`. Alternatively, we have to store the translations in `ResourceBundles` without the ability to nicely translate them.

**Comment by alexbakker** [comment]
Apologies for the wait here. I'll take a look at how we can do translations for this tonight. If we can't come up with a good solution, we'll go ahead and merge this without translations for now.

**Comment by orangenbaumblatt** [comment]
> Apologies for the wait here. I'll take a look at how we can do translations for this tonight. If we can't come up with a good solution, we'll go ahead and merge this without translations for now.

No problem. Let me just test it once again, I think a merged commit might resulted in some conflicts. I'll let you know if it's ready.

**Comment by alexbakker** [comment]
I can't really come up with a nice way to integrate ResourceBundles with Android's translation system either, so let's figure that out later. If need be, we can swap out zxcvbn4j for nbvcxz in the future, so that we have more translations for this out of the box. Shouldn't be too much work, as the two library's have similar API's.

**Comment by orangenbaumblatt** [comment]
I did a rebuild on the newest commit and fixed all of the conflicts.

**Comment by michaelschattgen** [comment]
Thank you for your work! We will review and merge this as soon as possible.

**Comment by alexbakker** [comment]
Thanks a lot!

**Pull request merged**